src/Illuminate/Database/Query/Builder.php

```diff
@@ -2703,20 +2723,57 @@ public function soleValue($column)
      */
     public function get($columns = ['*'])
     {
+        if (! is_null($keyBy = $this->keyBy)) {
+            return $this->getKeyed($keyBy, $columns);
+        }
+
         return collect($this->onceWithColumns(Arr::wrap($columns), function () {
             return $this->processor->processSelect($this, $this->runSelect());
         }));
     }
 
+    /**
+     * Execute the query as a "select" statement where the resulting array is keyed by the given column.
+     *
+     * @param  string  $key
+     * @param  array|string  $columns
+     * @return \Illuminate\Support\Collection
+     */
+    protected function getKeyed($key, $columns = ['*'])
+    {
+        return collect($this->onceWithColumns(Arr::prepend($this->qualifyStarColumns($columns), $key), function () {
```

Comment:
some explanation here: the column that you will use as the key will need to be prepended here since PDO will remove it in the process of assigning it as the key (caused by `PDO::FETCH_UNIQUE`). Furthermore, a star `'*'` has to be prefixed with the table name since `SELECT key_column, * FROM my_table` is not valid SQL (apparently putting `*` first IS valid SQL but that doesn't work in this case). The result then becomes `SELECT key_column, my_table.* FROM my_table`.